pkg/sql/distsql_physical_planner.go

```diff
@@ -957,23 +957,55 @@ func (h *distSQLNodeHealth) check(ctx context.Context, sqlInstanceID base.SQLIns
 func (dsp *DistSQLPlanner) PartitionSpans(
 	planCtx *PlanningCtx, spans roachpb.Spans,
 ) ([]SpanPartition, error) {
+	sp, _, err := dsp.PartitionSpansWithUserData(planCtx, spans, nil)
+	return sp, err
+}
+
+// PartitionSpansWithUserData splits the spans according to owning nodes just
+// like PartitionSpans, but also allows the spans to be associated with
+// arbitrary data. The data is duplicated whenever a span is split so that the
+// splits of a span is always associated with the original span's data.
+func (dsp *DistSQLPlanner) PartitionSpansWithUserData(
+	planCtx *PlanningCtx, spans []roachpb.Span, userData []interface{},
+) ([]SpanPartition, [][]interface{}, error) {
 	if len(spans) == 0 {
 		panic("no spans")
 	}
 	ctx := planCtx.ctx
 	partitions := make([]SpanPartition, 0, 1)
+	var userDataByPartition [][]interface{}
+
+	getUserData := func(i int) (interface{}, error) {
+		if userData == nil {
+			return nil, nil
+		}
+
+		if i < 0 || i >= len(userData) {
+			return nil, errors.Errorf("cannot get userData with index %d", i)
+		}
+		return userData[i], nil
+	}
```

Comment:
it feels like you may as well just check that userData and spans are the same size and then just use the same `i` value in the loop below. This feels contrived.

Reply:
Your other suggestion removes this code completely, so nothing more to do here.